### PR TITLE
Find NCCL2 debians in configure.py

### PR DESCRIPTION
**Pull request by jayfurmanek**
The Tensorlfow configure doesn't work with NCCL2 when it is installed via the debians.
This PR adds code that will handle the following scenarios:
 - NCCL2 installed via unpacked .tar file and not listed in ldconfig
 - NCCL2 installed via unpacked .tar file and listed properly in ldconfig
 - NCCL2 installed via debian file and not listed in ldconfig
 - NCCL2 installed via debian file and listed properly in ldconfig

The debian places the lib and header in the system paths (/usr/lib/*-linux-gnu and /usr/include) and the license files are placed in /usr/share/ and gzipped, so third_party/nccl/nccl_configure.bzl had to change in order to be flexible enough to find those, but changes here were deliberately kept to a minimum.

This avoids having to add a new config option for the NCCL2 header location by assuming /usr as the install location when NCCL is installed via debian.

### CODE REVIEW

**Comment by jayfurmanek** [comment]
Sorry -I was out. Looking at this now.

**Comment by jayfurmanek** [comment]
So the merge conflict was because some basic checking for the existence of the NCCL license was added recently. I added a check for the license for each of the scenarios listed in the description. I also reformatted a bit for consistency.

**Comment by jayfurmanek** [comment]
So, the license checking that cause the initial code conflict is now removed, causing another conflict. I'm inclined to leave it in for this PR as this is attempting to be comprehensive. I'd like some feedback, however. With the move of using NCCL2 as the default, I think we want more comprehensive support in configure.py.
Any thoughts/preferences before I resolve the conflict?

**Comment by jayfurmanek** [comment]
I resolved the conflict:
- I already had the expanduser calls in there so ~home dirs should work
- I left in the license checking

I think this PR improves NCCL2 configuration in configure.py, which is currently not that great.

**Review by gunan — DISMISSED** [review]

**Comment by jayfurmanek** [comment]
I can't see the internal build failures here. Let me know if there is anything I can help with here to get this one going.

**Comment by gunan** [comment]
I think the problem is due to nccl path not having a meaningful default?

**Comment by jayfurmanek** [comment]
Well, when looking for `libnccl.so.2` there are two scenarios:
 - ldconfig is properly set up and knows the location
 - ldconfig is not aware of nccl
If ldconifg knows the location, this PR just retrieves it and continues. There is no need to prompt the user and therefore no need for a default.
If ldconfig does not know the location, then the user is prompted and the default location is `CUDA_TOOLKIT_PATH` which should give us a good chance of finding libnccl assuming it was put in the proper location.

@gunan can you share any error messages from the internal builds?

**Comment by gunan** [comment]
The error messages I see are as early as configure script, unless we explicitly specify nccl locations, the configure script crashes and build fails. Currently, not sure how we resolve nccl location, but maybe @aaroey can weigh in.

**Comment by aaroey** [comment]
@tfboyd @smit-hinsu

File: configure.py [hunk 1]
```diff
@@ -1110,46 +1110,127 @@ def set_tf_nccl_install_path(environ_cp):
...
+        nccl_install_path = os.path.dirname(nccl2_path_from_ldconfig)
+        print('NCCL libraries found in ' + nccl2_path_from_ldconfig)
+        
+        # Check if this is the main Linux system lib location
+        # If so, mark the install root path as simply /usr
+        # so nccl_configure.bzl can find both the header and lib
+        if nccl_install_path.startswith("/usr/lib") and re.search('.*linux-gnu', nccl_install_path):
+          nccl_root_path = "/usr"
+          print("This looks like a system path. Using: " + nccl_root_path)
+          # Look for license
```
Comment (tfboyd) [inline configure.py]: I took out the part looking for the license due to the dynamic link and we do not include the license file in the build as users are not bound to it.

Comment (jayfurmanek) [reply configure.py]: Thanks @tfboyd ! So the unviewable internal build errors are caused by the inclusion of the search and collection of the license file?

>> we do not include the license file in the build as users are not bound to it.

I'll take your word for that. :)

**Comment by tfboyd** [comment]
I am not totally positive the license is the issue.  I suspect the issue is the following and I hope to make sometime to dig in or get Smit to do it.  :-)

-  Our build machine for normal builds has the .tgz install method and copied into /user/local/cuda or something like that or not and we depend on the ./configure hard coded dir structure.
-  Our Docker builds install with apt-get

**Comment by jayfurmanek** [comment]
The "proper" place for NCCL is to unpack it into the cuda directory like:
```/usr/local/cuda-9.2/targets/ppc64le-linux/lib/libnccl.so.2.2.13
/usr/local/cuda-9.2/targets/ppc64le-linux/lib/libnccl.so.2
/usr/local/cuda-9.2/targets/ppc64le-linux/lib/libnccl.so
/usr/local/cuda-9.2/targets/ppc64le-linux/lib/libnccl_static.a
/usr/local/cuda-9.2/targets/ppc64le-linux/include/nccl.h
```
or the corresponding `targets` locations on other platforms. This PR should find it there whether or not ldconfig knows where it is.  It also should find it if installed via apt-get, which should be:

```
/usr/include/nccl.h
/usr/lib/powerpc64le-linux-gnu/libnccl.so
/usr/lib/powerpc64le-linux-gnu/libnccl_static.a
/usr/lib/powerpc64le-linux-gnu/libnccl.so.2
/usr/lib/powerpc64le-linux-gnu/libnccl.so.2.2.12
```

On the other hand, this PR would fail if the NCCL-SLA file was not there (or removed from where it was unpacked/installed to). I just pushed a commit to remove the license checking, so we'll see.

**Comment by tfboyd** [comment]
It may have been fixed but the header files was ending up in something like /usr/local or lib or something/header/nccl.h.  I think on a call with NVIDIA they mentioned working to fix that as it was not helpful.  I am excited about this PR it may take some days to run through the scenarios and update the back-end build system.  I also like the move that the build system assumes the files are stored where installed by apt-get when possible.

**Comment by jayfurmanek** [comment]
It looks like it failed, so the license wasn't the issue. I can't see the results for the internal CI build fails.
@tfboyd are you able to post the error message from one of the builds? thx!

**Comment by gunan** [comment]
Here is the error: 
```
Please specify the location where NCCL 2 library is installed. Refer to README.md for more details. [Default is /usr/local/cuda]:

Assuming header is at /usr/local/cuda-9.0/include/nccl.h
Invalid path to NCCL 2 toolkit, /usr/local/cuda-9.0/lib64/libnccl.so.2 or /usr/local/cuda-9.0/include/nccl.h is not found. Please use the O/S agnostic package of NCCL 2
Please specify the NCCL version you want to use. If NCCL 2.2 is not installed, then you can use version 1.3 that can be fetched automatically but it may have worse performance with multiple GPUs. [Default is 2.2]: 

Please specify the location where NCCL 2 library is installed. Refer to README.md for more details. [Default is /usr/local/cuda]:

Assuming header is at /usr/local/cuda-9.0/include/nccl.h
Invalid path to NCCL 2 toolkit, /usr/local/cuda-9.0/lib64/libnccl.so.2 or /usr/local/cuda-9.0/include/nccl.h is not found. Please use the O/S agnostic package of NCCL 2
Traceback (most recent call last):
  File "configure.py", line 1654, in <module>
    main()
  File "configure.py", line 1593, in main
    set_tf_nccl_install_path(environ_cp)
  File "configure.py", line 1221, in set_tf_nccl_install_path
    _DEFAULT_PROMPT_ASK_ATTEMPTS)
__main__.UserInputError: Invalid TF_NCCL setting was provided 10 times in a row. Assuming to be a scripting mistake.
```

The build script assumes that nccl can be found without extra additions (I think) @aaroey I think you made the build script changes for nccl.
What is the mechanism to find where nccl is?

**Comment by tfboyd** [comment]
You will want to see my comment.  I updated all of the builds for NCCL2, it is a mess because we mixed tgz and apt-get and NVIDIA packaged NCCL oddly (put the header in /usr/local/header or something crazy like that) and maybe since has fixed it.  We can 100% make this PR work and it will be better.  It will just take some testing.  I am in the middle of a bunch of benchmark tests or I would have jumped in already.  

If you want to work through it yourself:
   * try to do a build where you unzip the package to a directory and have the ./configure still work if you point to the root directory where you unzipped that has /lib /header.  
   * Then try it with the apt-get install where it just finds it in the path.

**Comment by tfboyd** [comment]
@jayfurmanek   I see the logic this way, as a strawperson:

- Can I find it in the path, check ld or whatever
- If not user provides ./configure with the path which would either have to be individually to the .so and .h or to a parent directory that assumes a structure underneath like the .tgz.

I am very interested in other ideas.  For CUDA I can just point to another folder and off I go. While I would not want to build with multiple NCCL versions on a single system I would want to be able to do that.  That is where option #2 comes in and providing the path. Do you ask for the path to both .so and .h or a parent with a known structure.  Thoughts?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by tfboyd** [comment]
@smit-hinsu   I am sorry I dropped this item.  Can you pick it up?  The build process should work with NCCL 2.x installed from debian.  Make sure to try the latest package as I think NVIDIA has been making changes.  Since NCCL is only on linux we could also just look by default in the debian locations and then ask for the original folder where we expect library/header to exist as the fall back.   Jason may already being doing something similar or cooler I did not look closely.  Thankyou very much.

**Comment by tfboyd** [comment]
Also sorry to @jayfurmanek .  This will get sorted out.

**Comment by jayfurmanek** [comment]
I like the new build template files in `third_party/nccl`. They should help simplify this PR a bit.

**Comment by smit-hinsu** [comment]
@jayfurmanek Let us know when the tests are fixed and this PR is ready for review.

**Comment by jayfurmanek** [comment]
Will do. I'll be working on this today. I've had to change a bit to resolve the conflict. Mostly all for the better. I hope to push a commit today.

**Comment by jayfurmanek** [comment]
Dang, I updated my branch to resolve the merge conflicts and it now I have a ton of commits on PR. I'll close and reopen a new one and reference this.

**Comment by jayfurmanek** [comment]
replaced with https://github.com/tensorflow/tensorflow/pull/22519

**Pull request closed without merging**